[TensorExpr] Add CUDA codegen.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34234 [WIP] Changes in test/test_jit_fuser.py test/test_jit_fuser_te.py.
* #34233 [WIP] Changes to peephole.cpp.
* #34231 [WIP] Guard elimination changes from bertmaher/pytorch_fusion.
* #34230 [TensorExpr] Add tensorexpr benchmarks.
* #34228 [TensorExpr] Add LLVM codegen.
* **#34227 [TensorExpr] Add CUDA codegen.**
* #34226 [TensorExpr] Add a fuser pass based on tensor expressions.
* #34224 [TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch.

This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

Commit:
Add CUDA codegen for tensorexpressions.

[ghstack-poisoned]

Commit:
Update on "Add CUDA codegen for tensorexpressions."


Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update descriptions. on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Lints on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

resistor commented on torch/csrc/jit/tensorexpr/cuda_random.h:
```diff
@@ -0,0 +1,104 @@
+
+constexpr auto philox_random_string = R"(
+
+class Philox {
```
Would be nice if this could be shared with the eager mode philox implementation

Commit:
Rebase on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

bertmaher reviewed (approved):
Reviewed in side branch: http://github.com/bertmaher/pytorch/tree/pytorch_fusion/

Commit:
Rebase and try to fix windows build. on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add CUDA codegen."


This PR adds a CUDA support to tensor expressions.

Differential Revision: [D20251836](https://our.internmc.facebook.com/intern/diff/D20251836)

[ghstack-poisoned]
